### PR TITLE
Network Usage - Ingress Egress per interface and global usage

### PR DESCRIPTION
**Pull request by rita7lopes**
Hello,

I've added to the UI server stats tab the following:

 - Current Network usage, outlined by ingress and egress, with dynamic unit value so that every customer can see it regardless of the traffic amount
 - Network dashboard,
    - scale is dynamic, according to the max amount of traffic (top value is always 1 - 1 MB, 1 KB, 1 GB depending on the max observed)
    - it defaults to show "All" interfaces traffic summed, but the user can select the interface they want to observe and it persists on reload.
 
I've integrated the above into the implemented logic.

Also, I've tried merging main into my branch but the Dockerfile/deps to build are broken. I tried fixing but I failed. Thus, I used the files from the previous version, which work.

Images/video on the previous PR: https://github.com/mbecker20/komodo/pull/218

**WIP: Changing hashmap to vector type**

### CODE REVIEW

File: bin/periphery/src/api/mod.rs [hunk 1]
```diff
@@ -59,6 +59,11 @@ pub enum PeripheryRequest {
   #[to_string_resolver]
   GetSystemProcesses(GetSystemProcesses),
   GetLatestCommit(GetLatestCommit),
+  // Networks Stats (Read)
+  // #[to_string_resolver]
+  // GetNetworkStatsByInterface(GetNetworkStatsByInterface),
+  // #[to_string_resolver]
+  // GetTotalNetworkStats(GetTotalNetworkStats),
```
Comment (mbecker20) [inline bin/periphery/src/api/mod.rs:66]: These methods aren't needed, much like methods specifically to get disk usage aren't needed. This commented out stuff can be cleaned up.

File: bin/periphery/src/api/stats.rs [hunk 2]
```diff
@@ -56,3 +56,41 @@ impl ResolveToString<GetSystemProcesses> for State {
...
+//       _: (),
+//   ) -> anyhow::Result<String> {
+//       let client = stats_client().read().await;
+//       match client.get_network_stats_by_interface(&interface_name) {
+//         Some(stats) => serde_json::to_string(&stats)
+//             .context(format!("Failed to serialize network stats for interface: {}", interface_name)),
+//         None => Err(anyhow::anyhow!("Interface not found").into()),
+//     }
+//   }
+// }
```
Comment (mbecker20) [inline bin/periphery/src/api/stats.rs:96]: Can also be cleaned up

File: bin/periphery/src/stats.rs [hunk 3]
```diff
@@ -48,6 +48,29 @@ pub fn spawn_system_stats_polling_threads() {
...
+
+// /// Function to get the network ingress and egress data for a specific interface.
+// pub fn get_interface_network_usage(system: &System, interface_name: &str) -> Option<(f64, f64)> {
+//   let networks = system.networks();
+//   // Find the specified network interface and return its data
+//   networks.get(interface_name).map(|network_data| {
+//     (network_data.received(), network_data.transmitted())
+//   })
+// }
+
```
Comment (mbecker20) [inline bin/periphery/src/stats.rs:73]: Can also be cleaned up

File: bin/periphery/src/stats.rs [hunk 4]
```diff
@@ -57,6 +80,7 @@ pub struct StatsClient {
   // the handles used to get the stats
   system: sysinfo::System,
   disks: sysinfo::Disks,
+  network: sysinfo::Networks,
```
Comment (mbecker20) [inline bin/periphery/src/stats.rs:83]: rename field to `networks`, to match `disks` (and the `sysinfo::Networks` struct)

File: bin/periphery/src/stats.rs [hunk 5]
```diff
@@ -148,6 +208,23 @@ impl StatsClient {
...
+
+  // pub fn get_network_stats_by_interface(&self, interface_name: &str) -> Option<(f64, f64)> {
+  //   get_interface_network_usage(&self.system, interface_name).map(|(ingress, egress)| {
+  //     (
+  //         ingress as f64 / BYTES_PER_MB, // Convert to MB
+  //         egress as f64 / BYTES_PER_MB, // Convert to MB
+  //     )
+  //   })
+  // }
+
```
Comment (mbecker20) [inline bin/periphery/src/stats.rs:227]: cleanup commented

File: bin/periphery/src/stats.rs [hunk 6]
```diff
@@ -86,20 +112,54 @@ impl StatsClient {
...
+            SingleNetworkInterfaceUsage {
+                name: interface_name.clone(),
+                ingress_bytes: ingress,
+                egress_bytes: egress,
+            }
+        })
+        .collect();
+
+    // Debug Log the network usage interface to see if it's populated correctly
+    // println!("Network usage hash map: {:?}", network_usage);
```
Comment (mbecker20) [inline bin/periphery/src/stats.rs:151]: cleanup commented

File: client/periphery/rs/src/api/stats.rs [hunk 7]
```diff
@@ -18,8 +18,18 @@ pub struct GetSystemStats {}
...
-pub struct GetSystemProcesses {}
+// #[derive(Serialize, Deserialize, Debug, Clone, Request)]
+// #[response(NetworkStatsByInterface)]
+// pub struct GetNetworkStatsByInterface {}
+
+// //
+
+// #[derive(Serialize, Deserialize, Debug, Clone, Request)]
+// #[response(TotalNetworkStats)]
+// pub struct GetTotalNetworkStats {}
```
Comment (mbecker20) [inline client/periphery/rs/src/api/stats.rs:29]: cleanup commented

**Review by mbecker20 — CHANGES_REQUESTED** [review]
Hey, thanks for the PR. I like the idea of a selector for the interface, or view the total network. Lets start by cleaning this up a bit, and continue from there.

File: bin/periphery/src/stats.rs [hunk 8]
```diff
@@ -86,20 +112,54 @@ impl StatsClient {
...
   fn refresh_lists(&mut self) {
     self.disks.refresh_list();
+    self.network.refresh_list();
   }
 
   pub fn get_system_stats(&self) -> SystemStats {
     let total_mem = self.system.total_memory();
     let available_mem = self.system.available_memory();
+
+    let mut total_ingress: f64 = 0.0;
```
Comment (mbecker20) [inline bin/periphery/src/stats.rs:127]: The totals should use the numerical type returned by eg `network.received()`. The cast to f64 should happen only at the end.

File: bin/periphery/src/stats.rs [hunk 9]
```diff
@@ -86,20 +112,54 @@ impl StatsClient {
...
+    let mut total_egress: f64 = 0.0; 
+
+    // Fetch network data (Ingress and Egress)
+    let network_usage: Vec<SingleNetworkInterfaceUsage> = self.network
+        .iter()
+        .map(|(interface_name, stats)| {
```
Comment (mbecker20) [inline bin/periphery/src/stats.rs:133]: rename `stats` -> `network`, makes it clearer what we are looping through at a glance.

File: bin/periphery/src/api/mod.rs (hunk 1 above)
Comment (rita7lopes) [reply bin/periphery/src/api/mod.rs]: Done

File: bin/periphery/src/api/stats.rs (hunk 2 above)
Comment (rita7lopes) [reply bin/periphery/src/api/stats.rs]: Done

File: bin/periphery/src/stats.rs (hunk 3 above)
Comment (rita7lopes) [reply bin/periphery/src/stats.rs:51]: Done

File: bin/periphery/src/stats.rs (hunk 4 above)
Comment (rita7lopes) [reply bin/periphery/src/stats.rs]: Done

File: bin/periphery/src/stats.rs (hunk 5 above)
Comment (rita7lopes) [reply bin/periphery/src/stats.rs]: Done

File: bin/periphery/src/stats.rs (hunk 6 above)
Comment (rita7lopes) [reply bin/periphery/src/stats.rs]: Done

File: bin/periphery/src/stats.rs (hunk 8 above)
Comment (rita7lopes) [reply bin/periphery/src/stats.rs]: I changed to u64 and converted later to f64, but I don't know if this is what you had in mind. Please let me know

File: bin/periphery/src/stats.rs (hunk 9 above)
Comment (rita7lopes) [reply bin/periphery/src/stats.rs]: Done

File: bin/periphery/src/stats.rs [hunk 10]
```diff
@@ -86,20 +112,54 @@ impl StatsClient {
...
+    // Debug Log the network usage interface to see if it's populated correctly
+    // println!("Network usage hash map: {:?}", network_usage);
+
     SystemStats {
       cpu_perc: self.system.global_cpu_usage(),
       mem_free_gb: self.system.free_memory() as f64 / BYTES_PER_GB,
       mem_used_gb: (total_mem - available_mem) as f64 / BYTES_PER_GB,
       mem_total_gb: total_mem as f64 / BYTES_PER_GB,
+      // Added total ingress and egress
+      net_ingress_bytes: total_ingress as f64,
```
[the remark this answers is not in the document]
Comment (rita7lopes) [reply bin/periphery/src/stats.rs]: Done

File: bin/periphery/src/stats.rs (hunk 8 above)
Comment (mbecker20) [reply bin/periphery/src/stats.rs]: Perfect

**Review by mbecker20 — APPROVED** [review]
Looks good, I tested and it works! I'll merge to main and rebase 1.16.13 on top, so it will be included in the release. There may be some other tweaks I want to make before then. Thanks for your contribution!

**Pull request merged**